tests/test_async_utils.cc

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright (C) 2020 Swift Navigation Inc.
+ * Contact: Swift Navigation <dev@swiftnav.com>
+ *
+ * This source is subject to the license found in the file 'LICENSE' which must
+ * be distributed together with this source. All other rights reserved.
+ *
+ * THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND,
+ * EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
+ */
+
+#include <gtest/gtest.h>
+
+#include <albatross/Indexing>
+#include <albatross/utils/AsyncUtils>
+#include <albatross/utils/RandomUtils>
+
+#include <chrono>
+#include <mutex>
+#include <numeric>
+
+namespace albatross {
+
+TEST(test_async_utils, test_async_apply_with_capture) {
+  std::vector<int> xs = {0, 1, 2, 3, 4, 5};
+
+  std::mutex mu;
+  int sum = 0.;
+
+  std::vector<int> order_processed;
+
+  auto add_to_sum = [&](const int x) {
+    std::this_thread::sleep_for(std::chrono::milliseconds(abs(x - 2)));
+    mu.lock();
```

Comment:
It's preferred to use `std::lock_guard<std::mutex>` for locking a mutex.

Reply:
cool yeah that seems much nicer.